dataplane/internal/engine/engine.go

```diff
@@ -233,45 +282,249 @@ func (e *Engine) AddLayer3PuntRule(ctx context.Context, portName string, ip []by
+		Action: &fwdpb.ActionDesc_Select{
+			Select: &fwdpb.SelectActionListActionDesc{
+				SelectAlgorithm: fwdpb.SelectActionListActionDesc_SELECT_ALGORITHM_CRC32, // TODO: should algo + hash be configurable?
+				FieldIds: []*fwdpb.PacketFieldId{{Field: &fwdpb.PacketField{ // Hash the traffic flow, identified, IP protocol, L3 SRC, DST address, and L4 ports (if present).
+					FieldNum: fwdpb.PacketFieldNum_PACKET_FIELD_NUM_IP_PROTO,
+				}}, {Field: &fwdpb.PacketField{
+					FieldNum: fwdpb.PacketFieldNum_PACKET_FIELD_NUM_IP_ADDR_SRC,
+				}}, {Field: &fwdpb.PacketField{
+					FieldNum: fwdpb.PacketFieldNum_PACKET_FIELD_NUM_IP_ADDR_DST,
+				}}, {Field: &fwdpb.PacketField{
+					FieldNum: fwdpb.PacketFieldNum_PACKET_FIELD_NUM_L4_PORT_SRC,
+				}}, {Field: &fwdpb.PacketField{
+					FieldNum: fwdpb.PacketFieldNum_PACKET_FIELD_NUM_L4_PORT_DST,
+				}}},
+				ActionLists: actLists,
+			},
+		},
+	}, {
+		ActionType: fwdpb.ActionType_ACTION_TYPE_LOOKUP,
+		Action: &fwdpb.ActionDesc_Lookup{
+			Lookup: &fwdpb.LookupActionDesc{
+				TableId: &fwdpb.TableId{ObjectId: &fwdpb.ObjectId{
+					Id: nhTable,
+				}},
+			},
+		},
+	}}
+
+	entries := &fwdpb.TableEntryAddRequest{
+		ContextId: &fwdpb.ContextId{Id: e.id},
+		TableId: &fwdpb.TableId{
+			ObjectId: &fwdpb.ObjectId{
+				Id: nhgTable,
+			},
+		},
+		Entries: []*fwdpb.TableEntryAddRequest_Entry{{
+			EntryDesc: &fwdpb.EntryDesc{
+				Entry: &fwdpb.EntryDesc_Exact{
+					Exact: &fwdpb.ExactEntryDesc{
+						Fields: []*fwdpb.PacketFieldBytes{{
+							Bytes: binary.BigEndian.AppendUint64(nil, id),
+							FieldId: &fwdpb.PacketFieldId{
+								Field: &fwdpb.PacketField{
+									FieldNum: fwdpb.PacketFieldNum_PACKET_FIELD_NUM_NEXT_HOP_GROUP_ID,
+								},
+							},
+						}},
+					},
+				},
+			},
+			Actions: actions,
+		}},
+	}
+	if _, err := e.Server.TableEntryAdd(ctx, entries); err != nil {
+		return err
 	}
 
-	actions := nextHopToActions(nextHops[0])
+	return nil
+}
 
-	if len(nextHops) > 1 {
-		var actLists []*fwdpb.ActionList
-		for _, nh := range nextHops {
-			actLists = append(actLists, &fwdpb.ActionList{
-				Weight:  nh.GetWeight(),
-				Actions: nextHopToActions(nh),
-			})
+// addNextHop adds an entry to the next hop table.
+func (e *Engine) addNextHop(ctx context.Context, id uint64, nh *dpb.NextHop) error {
+	var nextHopIP []byte
+	if nhIPStr := nh.GetIp(); nhIPStr != "" {
+		nextHop := net.ParseIP(nhIPStr)
+		nextHopIP = nextHop.To4()
+		if nextHopIP == nil {
+			nextHopIP = nextHop.To16()
 		}
+	}
+	// Set the next hop IP in the packet's metadata.
+	nextHopAct := fwdbuilder.Action(fwdbuilder.UpdateAction(fwdpb.UpdateType_UPDATE_TYPE_SET, fwdpb.PacketFieldNum_PACKET_FIELD_NUM_NEXT_HOP_IP).WithValue(nextHopIP)).Build()
+	if nextHopIP == nil {
+		nextHopAct = fwdbuilder.Action(fwdbuilder.UpdateAction(fwdpb.UpdateType_UPDATE_TYPE_COPY, fwdpb.PacketFieldNum_PACKET_FIELD_NUM_NEXT_HOP_IP).WithFieldSrc(fwdpb.PacketFieldNum_PACKET_FIELD_NUM_IP_ADDR_DST)).Build()
+	}
```

Comment:
mention that if nexthop IP is not specified that the packet is treated as directly connected? This still doesn't quite make sense to me because it should be dropped or go to the default route right?

Reply:
agreed added todo remove this (was old behavior kept for compat with how sysrib calls this API)